src/order-truelength.c

```diff
@@ -0,0 +1,226 @@
+
+/*
+ * Construct a new `truelength_info`
+ *
+ * Pair with `PROTECT_TRUELENGTH_INFO()` in the caller
+ */
+struct truelength_info new_truelength_info(R_xlen_t max_size_alloc) {
+  struct truelength_info info;
+
+  info.strings = vctrs_shared_empty_chr;
+  info.lengths = vctrs_shared_empty_raw;
+  info.uniques = vctrs_shared_empty_chr;
+  info.sizes = vctrs_shared_empty_int;
+  info.sizes_aux = vctrs_shared_empty_int;
+
+  info.size_alloc = 0;
+  info.max_size_alloc = max_size_alloc;
+  info.size_used = 0;
+
+  info.max_string_size = 0;
+
+  info.reencode = false;
+
+  return info;
+}
+
+// -----------------------------------------------------------------------------
+
+/*
+ * Reset the truelengths of all unique strings captured in `strings` using
+ * the original truelengths in `lengths`.
+ *
+ * Will be called after each character data frame column is processed, and
+ * at the end of `chr_order()` for a single character vector.
+ */
+void truelength_reset(struct truelength_info* p_truelength_info) {
+  R_xlen_t size = p_truelength_info->size_used;
+
+  for (R_xlen_t i = 0; i < size; ++i) {
+    SEXP string = p_truelength_info->p_strings[i];
+    R_xlen_t length = p_truelength_info->p_lengths[i];
+
+    SET_TRUELENGTH(string, length);
+  }
+
+  // Also reset vector specific details
+  p_truelength_info->size_used = 0;
+  p_truelength_info->max_string_size = 0;
+  p_truelength_info->reencode = false;
+}
+
+// -----------------------------------------------------------------------------
+
+static void truelength_realloc(struct truelength_info* p_truelength_info);
+
+/*
+ * Saves a unique CHARSXP `x` along with its original truelength and
+ * its "size" (i.e the number of characters). Will be reset later with
+ * `truelength_reset()`.
+ */
+void truelength_save(struct truelength_info* p_truelength_info,
+                     SEXP x,
+                     R_xlen_t truelength,
+                     R_xlen_t size) {
+  // Reallocate as needed
+  if (p_truelength_info->size_used == p_truelength_info->size_alloc) {
+    truelength_realloc(p_truelength_info);
+  }
+
+  // Push `x` and `length`
+  p_truelength_info->p_strings[p_truelength_info->size_used] = x;
+  p_truelength_info->p_lengths[p_truelength_info->size_used] = truelength;
+  p_truelength_info->p_uniques[p_truelength_info->size_used] = x;
+  p_truelength_info->p_sizes[p_truelength_info->size_used] = size;
+
+  // Bump number of used slots
+  ++p_truelength_info->size_used;
+}
+
+// -----------------------------------------------------------------------------
+
+static R_xlen_t truelength_realloc_size(struct truelength_info* p_truelength_info);
+
+static SEXP truelength_chr_extend(const SEXP* p_x, R_xlen_t size_old, R_xlen_t size_new);
+static SEXP truelength_int_extend(const int* p_x, R_xlen_t size_old, R_xlen_t size_new);
+static SEXP truelength_lengths_extend(const R_xlen_t* p_lengths, R_xlen_t size_old, R_xlen_t size_new);
```

Comment:
For `p_chr_resize()`, I've also switched to using `SET_STRING_ELT()`, since this is what `Rf_xlengthgets()` uses https://github.com/wch/r-source/blob/5bd6e3ce1430374105ebf02101f9d55173496cfe/src/main/builtin.c#L891

Reply:
It should be safe to write to a `STRING_PTR()` here since all the CHARSXP are necessarily from an older (or equal) generation. But we'd mess up the refcounts which is ugly and it's more prudent to just go through the R API for writes. So I agree we have to accept the performance cost of writing lots of elements to character vectors.

Note that internal R code uses macros, not functions, so the cost is lower for them. When possible we should use `xlengthgets()`.